### ISSUE TITLE
[win] Slow Tree expand

### ISSUE THREAD

**Issue by jukzi** [eclipse-platform/eclipse.platform.swt#901]
For example when showing the quick-type hierarchy of `java.lang.Object `multiple seconds are spend in SWT expanding the tree.
most time is spend in `CallWindwProc` that follows each TreeItem.setExpanded()'s `OS.SendMessage (hwnd, OS.TVM_EXPAND`.
I don't get why it is that slow since `org.eclipse.jface.viewers.AbstractTreeViewer.inputChanged(Object, Object)` sets `tree.setRedraw(false);` before that.
Help wanted.
Screenshot from Sampling:
![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/51790620/8c0d1bb1-5fc0-46e3-bc5a-b7ec6385c580)

**Comment by vogella** [comment]
See also https://github.com/eclipse-platform/eclipse.platform.swt/issues/882

**Comment by SyntevoAlex** [comment]
If you make a snippet, I could profile it with native profiler.
I remember that `Tree` likes to calculate scrollbars even with `tree.setRedraw(false)`.
It's likely that the problem will go away if you expand items while parent is collapsed, and in the very end expand parent.

**Comment by jukzi** [comment]
> It's likely that the problem will go away if you expand items while parent is collapsed, and in the very end expand parent.

I don't have a snippet, but when i move setExpanded() after expanding the children, performance problem is gone. :-) And i don't see a visual difference. Thanks.

**Comment by laeubi** [comment]
> It's likely that the problem will go away if you expand items while parent is collapsed, and in the very end expand parent.

Could this be made somehow in the JFace impl to perform it in such a way by default?

**Comment by jukzi** [comment]
> Could this be made somehow in the JFace impl to perform it in such a way by default?

Absolutely. See my proposed PR. @SyntevoAlex After doing so the next hotspot (2seconds )is in `CallWindwProc` inside `org.eclipse.swt.widgets.Tree.createItem()` is there a trick against that too? It also happens while the tree's redraw is disabled.

**Comment by SyntevoAlex** [comment]
You might want to read code comments here
https://github.com/eclipse-platform/eclipse.platform.swt/blob/02c8a22116599a507db2dd424dd20cc778db64d4/bundles/org.eclipse.swt/Eclipse%20SWT/win32/org/eclipse/swt/widgets/TreeItem.java#L93-L99
and here
https://github.com/eclipse-platform/eclipse.platform.swt/blob/02c8a22116599a507db2dd424dd20cc778db64d4/bundles/org.eclipse.swt/Eclipse%20SWT/win32/org/eclipse/swt/widgets/TreeItem.java#L1694-L1698

**Comment by vogella** [comment]
cc @basilevs
